org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/commands/AssociateGrammarCommand.java

```diff
@@ -0,0 +1,143 @@
+import org.eclipse.lemminx.utils.StringUtils;
+import org.eclipse.lsp4j.ExecuteCommandParams;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+import org.xml.sax.Attributes;
+import org.xml.sax.SAXException;
+import org.xml.sax.helpers.DefaultHandler;
+
+/**
+ * Associate a grammar to a given XML command.
+ * 
+ * @author Angelo ZERR
+ *
+ */
+public class AssociateGrammarCommand extends AbstractDOMDocumentCommandHandler {
+
+	public static final String COMMAND_ID = "xml.associate.grammar.insert";
+
+	public AssociateGrammarCommand(IXMLDocumentProvider documentProvider) {
+		super(documentProvider);
+	}
+
+	public enum GrammarBindingType {
+
+		XSD("xsd"), //
+		DTD("dtd"), //
+		XML_MODEL("xml-model");
+
+		private String name;
+
+		private GrammarBindingType(String name) {
+			this.name = name != null ? name : name();
+		}
+
+		public String getName() {
+			return name;
+		}
+	}
+
+	@Override
+	protected Object executeCommand(DOMDocument document, ExecuteCommandParams params, SharedSettings sharedSettings,
+			CancelChecker cancelChecker) throws Exception {
+		String documentURI = document.getDocumentURI();
+		String fullPathGrammarURI = ArgumentsUtils.getArgAt(params, 1, String.class);
+		String bindingType = ArgumentsUtils.getArgAt(params, 2, String.class);
+		String grammarURI = getRelativeURI(fullPathGrammarURI, documentURI);
+
+		if (GrammarBindingType.XSD.getName().equals(bindingType)) {
+			// Check if XSD to bind declares a target namespace
+			String targetNamespace = getTargetNamespace(fullPathGrammarURI);
+			if (StringUtils.isEmpty(targetNamespace)) {
+				// Insert inside <foo /> ->
+				// xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance"
+				// xsi:noNamespaceSchemaLocation=\"xsd/tag.xsd\"
+				return NoGrammarConstraintsCodeAction.createXSINoNamespaceSchemaLocationEdit(grammarURI, document);
+			}
+			// Insert inside <foo /> ->
+			// xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+			// xsi:noNamespaceSchemaLocation="xsd/tag.xsd"
+			return NoGrammarConstraintsCodeAction.createXSISchemaLocationEdit(grammarURI, targetNamespace, document);
+		} else if (GrammarBindingType.DTD.getName().equals(bindingType)) {
+			// Insert before <foo /> -> <!DOCTYPE foo SYSTEM "dtd/tag.dtd">
+			return NoGrammarConstraintsCodeAction.createDocTypeEdit(grammarURI, document, sharedSettings);
+		} else if (GrammarBindingType.XML_MODEL.getName().equals(bindingType)) {
+			// Insert before <foo /> -> <?xml-model href=\"dtd/tag.dtd\"?>
+			return NoGrammarConstraintsCodeAction.createXmlModelEdit(grammarURI, document, sharedSettings);
+		}
+		return null;
+	}
+
+	private static String getRelativeURI(String fullPathGrammarURI, String documentURI) {
+		try {
+			Path grammarPath = FilesUtils.getPath(fullPathGrammarURI);
+			Path documentPath = FilesUtils.getPath(documentURI);
+			Path relativePath = documentPath.getParent().relativize(grammarPath);
+			return relativePath.toString().replaceAll("[\\\\]", "/");

```

Comment:
The regex `[\\]` and `\\` both match a single `\`, so I think that you could replace `"[\\\\]"` with `"\\\\"`.

Reply:
Good catch! Fixed